packages/transformers/js/src/JSTransformer.js

```diff
@@ -152,6 +152,90 @@ type TSConfig = {
 };
 
 export default (new Transformer({
+  async loadGlobalConfig({config, options}) {
+    let result = await config.getConfigFrom<PackageJSONConfig>(
+      path.join(options.projectRoot, 'index'),
+      ['package.json'],
+    );
+    let rootPkg = result?.contents;
+
+    let inlineEnvironment;
+    let inlineFS = true;
+    if (result && rootPkg?.['@parcel/transformer-js']) {
+      validateSchema.diagnostic(
+        CONFIG_SCHEMA,
+        {
+          data: rootPkg['@parcel/transformer-js'],
+          // FIXME
+          source: await options.inputFS.readFile(result.filePath, 'utf8'),
+          filePath: result.filePath,
+          prependKey: `/${encodeJSONKeyComponent('@parcel/transformer-js')}`,
+        },
+        // FIXME
+        '@parcel/transformer-js',
+        'Invalid config for @parcel/transformer-js',
+      );
+
+      inlineEnvironment = rootPkg['@parcel/transformer-js']?.inlineEnvironment;
+      inlineFS = rootPkg['@parcel/transformer-js']?.inlineFS ?? inlineFS;
+    }
+
+    // Injected into source and external files
+    let publicEnv: EnvMap = {};
+    // Injected into only source files
+    let privateEnv: EnvMap = {};
+
+    let assignAllEnv = (env: EnvMap) => {
+      for (let key in options.env) {
+        if (!key.startsWith('npm_')) {
+          env[key] = String(options.env[key]);
+        }
+      }
+    };
+
+    let assignDefaultEnv = (env: EnvMap) => {
+      if (options.env.NODE_ENV != null) {
+        env.NODE_ENV = options.env.NODE_ENV;
+      }
+
+      if (process.env.PARCEL_BUILD_ENV === 'test') {
+        env.PARCEL_BUILD_ENV = 'test';
+      }
+    };
+
+    if (inlineEnvironment == null) {
+      // Default behaviour
+      // Source files can inject all ENV vars
+      // External files can only inject `NODE_ENV` and `PARCEL_BUILD_ENV`
+      assignDefaultEnv(publicEnv);
+      assignAllEnv(privateEnv);
+    } else if (Array.isArray(inlineEnvironment)) {
+      // inlineEnvironment allowlist
+      // All files can inject ENV vars in allow list
+      for (let match of globMatch(
+        Object.keys(options.env),
+        inlineEnvironment,
+      )) {
+        publicEnv[match] = String(options.env[match]);
+      }
+    } else if (inlineEnvironment === true) {
+      // inlineEnvironment: true
+      // All files can inject all ENV vars
+      assignAllEnv(publicEnv);
+    } else if (inlineEnvironment === false) {
```

Comment:
While I think the existing `inlineEnvironment` behaviour is a bit confusing, this is attempting to replicate it exactly. 

Specifically, I'm not a huge fan of default behaviour being different to all the possible config options. We can discuss the actual behaviour separately though.